proposals/Exceptions.md

```diff
@@ -143,42 +93,85 @@ Exception indices are used by:
 following form:
 
 ```
 try blocktype
   instruction*
-catch
+catch i
+  instruction*
+catch j
+  instruction*
+...
+catch_all
   instruction*
 end
 ```
 
-A try block ends with a `catch block` that is defined by the list of
-instructions after the `catch` instruction.
+A try-catch block contains zero or more `catch` blocks and zero or one
+`catch_all` block. All `catch` blocks should precede the `catch_all` block, if
+any. The `catch`/`catch_all` instructions (within the try construct) are called
+the _catching_ instructions. There should be at least `catch` or `catch_all`
+block within a try-catch block.
+
+The _body_ of the try block is the list of instructions before the first
+catching instruction. The _body_ of each catch block is the sequence of
+instructions following the corresponding catching instruction before the next
+catching instruction (or the `end` instruction if it is the last catching
+block).
+
+The `catch` instruction has an exception tag associated with it. The tag
+identifies what exceptions it can catch. That is, any exception created with the
+corresponding exception tag. Catch blocks that begin with a `catch` instruction
+are considered _tagged_ catch blocks.
+
+The last catching instruction of a try-catch block can be the `catch_all`
+instruction. If it begins with the `catch_all` instruction, it defines the
+_default_ catch block. The default catch block has no exception type, and is
+used to catch all exceptions not caught by any of the tagged catch blocks. The
+term 'catch block' refers to both `catch` and `catch_all` blocks.
 
 Try blocks, like control-flow blocks, have a _block type_. The block type of a
 try block defines the values yielded by evaluating the try block when either no
 exception is thrown, or the exception is successfully caught by the catch block.
 Because `try` and `end` instructions define a control-flow block, they can be
 targets for branches (`br` and `br_if`) as well.
 
-In the initial implementation, try blocks may only yield 0 or 1 values.
+### Try-unwind blocks
+
+Try blocks can also be used with the `unwind` instruction. A try-unwind block
+contains an `unwind` block with the following form:
+
+```
+try blocktype
+  instruction*
+unwind
+  instruction*
+end
+```
+
+The `unwind` block is meant to contain cleanup instructions, such as
+destructors, in case any instruction in the corresponding try block throws.
+Currently the `unwind` instruction has the same semantics as the `catch_all`
+instruction that it catches all exceptions. In case an exception is caught by
+the `unwind` block, it becomes the catching block.
+
+The `end` instruction at the end of `unwind` block is special that it
+automatically rethrows the current exception.
+
+The current plan is, try-unwind blocks will have a different semantics from
+that of try-catch blocks when we introduce two-phase unwinding as a follow-on
+proposal in the future.
```

Comment:
It would really help if this plan could be sketched in this explainer somehow (or some separate doc to point to). Because to be honest, I still don't understand it, nor how the current semantics of `unwind` could be changed/generalised later in a backwards-compatible manner.